File: 프로그래머스/2/87946. 피로도/피로도.py
from itertools import permutations


def solution(k, dungeons):
    answer = -1

    for check in permutations(range(len(dungeons))):
        count = 0
        current_k = k

        for seq in check:
            condition, cost = dungeons[seq]
            
            if current_k < condition:
                break
                
            current_k -= cost
            count += 1

        answer = max(answer, count)

    return answer

File: 프로그래머스/2/87946. 피로도/피로도.java
class Solution {
    
    private int answer = -1;
    private boolean[] visited;

    public int solution(int k, int[][] dungeons) {
        visited = new boolean[dungeons.length];
        dfs(k, dungeons, 0);
        return answer;
    }

    private void dfs(int k, int[][] dungeons, int count) {
        answer = Math.max(answer, count);

        for (int i = 0; i < dungeons.length; i++) {
            int condition = dungeons[i][0];
            int cost = dungeons[i][1];

            if (!visited[i] && k >= condition) {
                visited[i] = true;
                dfs(k - cost, dungeons, count + 1);
                visited[i] = false;
            }
        }
    }
}